Support TLS certificate auto-generation using certmanager (#17238)
cilium-team/cilium: Documentation/concepts/observability/hubble-configuration.rst

.. _hubble_configure_tls_certs:

Use custom TLS certificates
---------------------------
TLS certificates
================

When Hubble Relay is deployed, Hubble listens on a TCP port on the host
network. This allows Hubble Relay to communicate with all Hubble instances in
the cluster. Connections between Hubble server and Hubble Relay instances are
secured using mutual TLS (mTLS) by default.

TLS certificates are automatically generated and distributed as Kubernetes
secrets for use by Hubble and Hubble Relay provided that
``hubble.tls.auto.enabled`` is set to ``true`` (default). This is achieved
either via Helm when using Helm or via a Kubernetes CronJob when using

> ``experimental-install.yaml``.

Reviewer: good catch, we don't have `experimental-install.yaml` anymore.

> via a Kubernetes CronJob or via cert-manager.

Reviewer: nitpicking: here is a rewording suggestion attempting to make it clearer as there are many options

```suggestion
TLS certificates can be provided by manually on the Helm install command (user
provided) or generate automatically via either:

* Helm
* certgen (a Kubernetes CronJob)
* cert-manager
```

TLS certificates are provided by user or automatically generated via Helm,
via a Kubernetes CronJob or via cert-manager.

.. note::

   When using Helm, TLS certificates are (re-)generated every time Helm is used
   for install or upgrade. As Hubble server and Hubble Relay support TLS
   certificates hot reloading, including CA certificates, this does not disrupt
   any existing connection. New connections are automatically established using
   the new certificates without having to restart Hubble server or Hubble
   Relay.

Hubble allows using custom TLS certificates rather than relying on
automatically generated ones. This can be useful when using Hubble Relay in a
cluster mesh scenario for instance or when using certificates signed by a
specific certificate authority (CA) is required.
User provided certificates
--------------------------

In order to use custom TLS certificates ``hubble.tls.auto.enabled`` must

> be set to ``false`` and TLS certificates manually provided.

Reviewer: I would rather keep this paragraph.

Author: It's back

Reviewer: Thanks!

--------------------------------------------------

Like helm method, TLS certificates are generated at installation time. And a kubernetes CronJob
is used to schedule for certificates regeneration (regardless of their expiration date).

.. code-block:: bash

    --set hubble.tls.auto.enabled=true               # enable automatic TLS certificate generation
    --set hubble.tls.auto.method=cronJob             # auto generate certificates using cronJob method
    --set hubble.tls.auto.certValidityDuration=1095  # certificates validity duration in days (default 3 years)
    --set hubble.tls.auto.schedule="0 0 1 */4 *"     # schedule for certificates re-generation (crontab syntax)

Auto generated certificates via cert-manager
--------------------------------------------

This method rely on `cert-manager <https://cert-manager.io/>`_ to generated certificate.
``cert-manager`` now become de-facto way to manage TLS on k8s, and compared to above methods,
it has the following advantages:

* No need to extra cronJob
* Support multiple issuers: CA, Vault, Let's Encrypt, Google CAS,...
  You can choose the issuer that fits your organization's requirements.

>   Note that ``helm`` and ``cronJob`` methods only generate self-signed certificates.

Reviewer: This is not technically correct as both Helm and certgen both generate a CA.

* Manage certs via a CRD, which is more straightforward than inspect the PEM file.
* Auto-renew certificates

**Installation steps**:

1. First install `cert-manager <https://cert-manager.io/docs/installation/>`_ and setup `issuer <https://cert-manager.io/docs/configuration/>`_.

>   Please make sure that your issuer can be able to create ``cilium.io`` domain certificate.

Reviewer: Clarify that subdomains are needed

```suggestion
  Please make sure that your issuer is be able to create certificates under the ``cilium.io`` domain name.
```

2. Install/upgrade cilium with bellow configs:

.. code-block:: bash

    --set hubble.tls.auto.enabled=true               # enable automatic TLS certificate generation
    --set hubble.tls.auto.method=certmanager         # auto generate certificates using cronJob method
    --set hubble.tls.auto.certValidityDuration=1095  # certificates validity duration in days (default 3 years)
    --set hubble.tls.auto.certManagerIssuerRef.group="cert-manager.io" # Reference to cert-manager's issuer
    --set hubble.tls.auto.certManagerIssuerRef.kind="ClusterIssuer"
    --set hubble.tls.auto.certManagerIssuerRef.name="ca-issuer"

**Troubleshooting**:

If you get the following error while install cilium (and cert-manager), it's because

> cert-manager ValidatingWebhook (which used to vefiry your CRD resourcs) is not available

Reviewer: typos

```suggestion
cert-manager ValidatingWebhook (which used to verify the Certificate CRD resources) is not available
```

(blocked due to CNI is not available).

.. code-block:: bash

    Error: Internal error occurred: failed calling webhook "webhook.cert-manager.io": Post "https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=10s": dial tcp x.x.x.x:443: connect: connection refused

There are several ways to overcome above issue, such as:

> * Option 1: disable validation on cilium installed namespace

Reviewer: A newline is required before the first item of list, otherwise it doesn't render correctly

```suggestion
There are several ways to workaround this issue:

* Option 1: disable validation on cilium installed namespace
```

.. code-block:: bash

    # We assum cilium installed in kube-system namespace
    kubectl label namespace kube-system cert-manager.io/disable-validation=true

    helm install cert-manager ...
    kubectl apply -f issuer.yaml
    helm install cilium ...

* Option 2: install cert-manager CRDs first

.. code-block:: bash

>     kubectl create -f https://github.com/jetstack/cert-manager/releases/download/v1.5.3/cert-manager.crds.yaml

Reviewer: We should avoid to hardcode the cert-manager version here, maybe link to https://cert-manager.io/docs/installation/helm/#option-1-installing-crds-with-kubectl instead.

--------------------------------------------------

Like helm method, TLS certificates are generated at installation time. And a kubernetes CronJob
is used to schedule for certificates regeneration (regardless of their expiration date).

.. code-block:: bash

    --set hubble.tls.auto.enabled=true               # enable automatic TLS certificate generation
    --set hubble.tls.auto.method=cronJob             # auto generate certificates using cronJob method
    --set hubble.tls.auto.certValidityDuration=1095  # certificates validity duration in days (default 3 years)
    --set hubble.tls.auto.schedule="0 0 1 */4 *"     # schedule for certificates re-generation (crontab syntax)

Auto generated certificates via cert-manager
--------------------------------------------

This method rely on `cert-manager <https://cert-manager.io/>`_ to generated certificate.
``cert-manager`` now become de-facto way to manage TLS on k8s, and compared to above methods,
it has the following advantages:

* No need to extra cronJob
* Support multiple issuers: CA, Vault, Let's Encrypt, Google CAS,...
  You can choose the issuer that fits your organization's requirements.

> * Manage certs via a CRD, which is more straightforward than inspect the PEM file.

Reviewer 2: ```suggestion
* Manage certs via a CRD, which is more straightforward than inspecting the PEM file.
```

    --set hubble.tls.auto.schedule="0 0 1 */4 *"     # schedule for certificates re-generation (crontab syntax)

Auto generated certificates via cert-manager
--------------------------------------------

This method rely on `cert-manager <https://cert-manager.io/>`__ to generated certificate.
``cert-manager`` now become de-facto way to manage TLS on k8s, and compared to above methods,
it has the following advantages:

* No need to extra cronJob
* Support multiple issuers: CA, Vault, Let's Encrypt, Google CAS,...
  You can choose the issuer that fits your organization's requirements.
* Manage certs via a CRD, which is more straightforward than inspecting the PEM file.
* Auto-renew certificates

**Installation steps**:

1. First install `cert-manager <https://cert-manager.io/docs/installation/>`__ and setup `issuer <https://cert-manager.io/docs/configuration/>`_.
   Please make sure that your issuer is be able to create certificates under the ``cilium.io`` domain name.
2. Install/upgrade cilium with bellow configs:

.. code-block:: bash

    --set hubble.tls.auto.enabled=true               # enable automatic TLS certificate generation

>     --set hubble.tls.auto.method=certmanager         # auto generate certificates using cronJob method

Reviewer 3: ```suggestion
    --set hubble.tls.auto.method=certmanager         # auto generate certificates using cert-manager
```

    kubectl label namespace kube-system cert-manager.io/disable-validation=true

    helm install cert-manager ...
    kubectl apply -f issuer.yaml
    helm install cilium ...

* Option 2: install cert-manager CRDs first

.. code-block:: bash

    # see https://cert-manager.io/docs/installation/helm/#option-1-installing-crds-with-kubectl
    kubectl create -f cert-manager.crds.yaml

    # cert-manager MUST be installed after cilium
    helm install cilium ...
    
    helm install cert-manager ...
    kubectl apply -f issuer.yaml

* Option 3: install cert-manager webhook with hostNetwork

.. code-block:: bash

    helm install cert-manager jetstack/cert-manager \
            --set webhook.hostNetwork=true \
            --set webhook.tolerations='["operator": "Exists"]'
    kubectl apply -f issuer.yaml

    helm install cilium ...

* Option 4: upgrade cilium from disabled-TLS installation

.. code-block:: bash

    helm install cilium cilium/cilium \
            --set hubble.tls.enabled=false \
            ...
    
    helm install cert-manager ...
    kubectl apply -f issuer.yaml

    # waiting for node ready, and cert-manager available

>     helm upgrade cilium ...

Reviewer 3: nit: Explicitly use `--set hubble.tls.enabled=true` here to make it clear that we want to enable TLS in this step.

   of the certificate for Hubble server MUST be set to
   ``*.{cluster-name}.hubble-grpc.cilium.io`` where ``{cluster-name}`` is the
   cluster name defined by ``cluster.name`` (defaults to ``default``).

Auto generated certificates via Helm
------------------------------------

When using Helm, TLS certificates are (re-)generated every time Helm is used
for install or upgrade. As Hubble server and Hubble Relay support TLS
certificates hot reloading, including CA certificates, this does not disrupt
any existing connection. New connections are automatically established using
the new certificates without having to restart Hubble server or Hubble
Relay.

> .. code-block:: bash

Reviewer 4: Sorry, missed this last time - none of these code blocks are bash. I would suggest replacing each of the `.. code-block:: bash` with `::` instead, since they're just text snippets.
```suggestion
::
```

.. code-block:: bash

    --set hubble.tls.auto.enabled=true               # enable automatic TLS certificate generation
    --set hubble.tls.auto.method=certmanager         # auto generate certificates using cert-manager
    --set hubble.tls.auto.certValidityDuration=1095  # certificates validity duration in days (default 3 years)
    --set hubble.tls.auto.certManagerIssuerRef.group="cert-manager.io" # Reference to cert-manager's issuer
    --set hubble.tls.auto.certManagerIssuerRef.kind="ClusterIssuer"
    --set hubble.tls.auto.certManagerIssuerRef.name="ca-issuer"

**Troubleshooting**:

If you get the following error while install cilium (and cert-manager), it's because
cert-manager ValidatingWebhook (which used to verify the Certificate CRD resources)
is not available (blocked due to CNI is not available).

.. code-block:: bash

    Error: Internal error occurred: failed calling webhook "webhook.cert-manager.io": Post "https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=10s": dial tcp x.x.x.x:443: connect: connection refused

There are several ways to overcome above issue, such as:

* Option 1: disable validation on cilium installed namespace

> .. code-block:: bash
> 
>     # We assum cilium installed in kube-system namespace
>     kubectl label namespace kube-system cert-manager.io/disable-validation=true
> 
>     helm install cert-manager ...
>     kubectl apply -f issuer.yaml
>     helm install cilium ...

Reviewer 4: For these ones, if you want to format them as a shell-session then you need to do:
```suggestion
.. code-block:: shell-session

    $ # We assume cilium installed in kube-system namespace
    $ kubectl label namespace kube-system cert-manager.io/disable-validation=true

    $ helm install cert-manager ...
    $ kubectl apply -f issuer.yaml
    $ helm install cilium ...
```

Reviewer 4: Also, looking at these snippets, I see `issuer.yaml` mentioned but I'm not sure where that comes from. How does the user generate `issuer.yaml`?

Author: @joestringer I also add a note that where `issuer.yaml` is come from.

   of the certificate for Hubble server MUST be set to
   ``*.{cluster-name}.hubble-grpc.cilium.io`` where ``{cluster-name}`` is the
   cluster name defined by ``cluster.name`` (defaults to ``default``).

Auto generated certificates via Helm
------------------------------------

When using Helm, TLS certificates are (re-)generated every time Helm is used
for install or upgrade. As Hubble server and Hubble Relay support TLS
certificates hot reloading, including CA certificates, this does not disrupt
any existing connection. New connections are automatically established using
the new certificates without having to restart Hubble server or Hubble
Relay.

::

    --set hubble.tls.auto.enabled=true               # enable automatic TLS certificate generation
    --set hubble.tls.auto.method=helm                # auto generate certificates using helm method
    --set hubble.tls.auto.certValidityDuration=1095  # certificates validity duration in days (default 3 years)

> The downside of this method is certificates is not auto-renew.

Reviewer 5: ```suggestion
The downside of this method is certificates are not auto-renewed.
```

So re-install is required when certificates are expired.

Auto generated certificates via Kubernetes CronJob
--------------------------------------------------

Like helm method, TLS certificates are generated at installation time. And a kubernetes CronJob
is used to schedule for certificates regeneration (regardless of their expiration date).

::

    --set hubble.tls.auto.enabled=true               # enable automatic TLS certificate generation
    --set hubble.tls.auto.method=cronJob             # auto generate certificates using cronJob method
    --set hubble.tls.auto.certValidityDuration=1095  # certificates validity duration in days (default 3 years)
    --set hubble.tls.auto.schedule="0 0 1 */4 *"     # schedule for certificates re-generation (crontab syntax)

Auto generated certificates via cert-manager
--------------------------------------------

> This method rely on `cert-manager <https://cert-manager.io/>`__ to generated certificate.

Reviewer 5: ```suggestion
This method rely on `cert-manager <https://cert-manager.io/>`__ to generate certificate.
```

> ``cert-manager`` now become de-facto way to manage TLS on k8s, and compared to above methods,

Reviewer 5: ```suggestion
``cert-manager`` now becomes de-facto way to manage TLS on k8s, and compared to above methods,
```

it has the following advantages:

> * No need to extra cronJob

Reviewer 5: ```suggestion
* No need for extra cronJob
```
